File: android/src/main/java/com/rnemittermodule/HybridRnEmitterModule.kt
package com.rnemittermodule

import android.annotation.SuppressLint
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import com.margelo.nitro.NitroModules
import com.margelo.nitro.rnemittermodule.HybridRnEmitterModuleSpec

typealias RNMessageCallback = (msg: String, data: String?) -> Unit

data class RNListener<T>(
    val id: Double,
    val callback: T
)

@SuppressLint("UnspecifiedRegisterReceiverFlag")
open class HybridRnEmitterModule: HybridRnEmitterModuleSpec() {
    private val applicationContext = NitroModules.applicationContext?.applicationContext
    private val packageName: String? = applicationContext?.packageName
    private var currentListenerId = 0.0;
    private val listeners = mutableListOf<RNListener<RNMessageCallback>>()

    private val receiver = object : BroadcastReceiver() {
        override fun onReceive(ctx: Context?, intent: Intent?) {
            val event = intent?.getStringExtra("event") ?: return
            val data = intent.getStringExtra("data")

            listeners.forEach { it.callback(event, data) }
        }
    }

    init {
        packageName?.let {
            val filter = IntentFilter("$it.RNEmitterResponse")
            applicationContext?.registerReceiver(receiver, filter)
        }
    }

    override fun emitToNative(message: String, data: String?) {
        val action = packageName?.let { "$it.RNEmitterSend" } ?: return

        val intent = Intent(action).apply {
            putExtra("event", message)
            putExtra("data", data)
        }

        applicationContext?.sendBroadcast(intent)
    }

    override fun addNativeEventListener(callback: (event: String, data: String?) -> Unit): Double {
        currentListenerId += 1
        listeners.add(RNListener(currentListenerId, callback))
        return currentListenerId
    }

    override fun removeNativeEventListener(id: Double) {
        listeners.removeAll { it.id == id }
    }

    protected fun destroy() {
        applicationContext?.unregisterReceiver(receiver)
    }
}

File: ios/HybridRnEmitterModule.swift
import Foundation
import NitroModules

struct RNListener<T> {
    let id: Double
    let callback: T
}


typealias RNMessageCallback = (_ msg: String, _ data: String?) -> Void

class HybridRnEmitterModule: HybridRnEmitterModuleSpec {
    private var currentListenerId: Double = 0
    private var listeners: [RNListener<RNMessageCallback>] = []
    
    override init() {
        super.init()

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleDidRespondToReactNative(_:)),
            name: NSNotification.Name("RNEmitterResponse"),
            object: nil
        )
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    func emitToNative(message: String, data: String?) throws -> Void {
        
        let parsedData: [String: Any] = {
            guard
                let json = data?.data(using: .utf8),
                let obj = try? JSONSerialization.jsonObject(with: json),
                let dict = obj as? [String: Any]
            else {
                return [:]
            }
            
            return dict
        }()

        NotificationCenter.default.post(
            name: NSNotification.Name("RNEmitterSend"),
            object: nil,
            userInfo: [
                "event": message,
                "data": parsedData
            ]
        )
    }
    
    func addNativeEventListener(callback: @escaping (String, String?) -> Void) throws -> Double {
        currentListenerId += 1
        let listener = RNListener(id: currentListenerId, callback: callback)
        listeners.append(listener)
        return currentListenerId
    }
    
    func removeNativeEventListener(id: Double) throws -> Void {
        listeners.removeAll { $0.id == id }
    }
    
    @objc private func handleDidRespondToReactNative(_ notification: Notification) -> Void {
        guard let userInfo = notification.userInfo,
              let event = userInfo["event"] as? String else {
            return
        }
        
        
        let dataJson: String? = {
            guard let data = userInfo["data"] as? [String: Any],
                  let jsonData = try? JSONSerialization.data(withJSONObject: data, options: []) else {
                return nil
            }
            return String(data: jsonData, encoding: .utf8)
        }()
        
        

        listeners.forEach { $0.callback(event, dataJson) }
    }
}
